### DISCUSSION TITLE
Install Microsoft.EntityFrameworkCore.Tools to avoid embarrassment

### DISCUSSION THREAD

**Discussion by DiamondImpostor** [JetBrains/rider-efcore#113] [Suggestions]
It took me a while to figure out somewhere the plugin worked, but in some project it did not and why it wasn't looking for the Startup Project, until I found that you have to use the Tools package to migrate. It would be cool to offer to install this package somehow, or install it automatically in the project which contains DatabaseContext

Unfortunately I don't know kotlin and Jetbrains plugins code, I can't create a PR, but I believe it's easy to do, thank you

**Comment by seclerp** [comment]
Hello,

Basic concepts and rules of using EF Core are out of the scope of this plugin, Currently, it's a clever wrapper over `dotnet ef` command line tool, so all requirements that are suitable for `dotnet ef` are also suitable for the plugin.

You can't execute any `dotnet ef` command without having `.Design` or `.Tools` NuGet package installed into startup project.

But even if we go this way of adding the functionality that you are talking about, how exactly do you expect it to be presented on the UI? We have a dropdown for startup projects, only correct projects could be listed there. We could make a comment near this field to describe a requirement for the dropdown a bit better, for example.

**Comment by DiamondImpostor** [comment]
Hi, @seclerp, thanks for the detailed reply.

I also saw that some plugin users created #issues(https://github.com/seclerp/rider-efcore/issues/112) with the same question, and they were asked to install the Tools package, which is correct. But we can simplify it.

The simplest way to detect detect is Tools installed:
1) The plugin already finds DatabaseContext whether Tools is installed or not
2) The plugin already finds any projects to create Migrations whether Tools is installed or not
3) If the startup project is not found (the startup project is determined by the Tools installation definition, I think):
    ask user to install Tools package (like "Drop Database" confirmation dialogue)
    ![Image alt](https://i.imgur.com/N8Qm23p.jpeg)

The hard way is install EntityFrameworkCore.Tools for a DatabaseContext project, I don't think that will work for everyone. You can use "Tools package is missed" dialog, but use "Install" and "Cancel" instead of "Ok".

**Reply by seclerp** [reply]
The issue with that approach is that "Startup project" is not a project where your `DbContext` lives. Such a project is called "Migrations project".

**Migrations project** (usually) should not contain a reference to any tools NuGet packages, `Microsoft.EntityFrameworkCore` is usually enough for successful build and usage. In simple solutions, your **Startup** and **Migrations** projects are the same, but in a majority of real-world cases, they are not.

The only way to detect if a project is a valid candidate to be a **Startup project** in your suggestion is to collect projects those reference **Migrations project** and to try to find it inside this collected list. But with that in mind, you will receive a lot of false-positive suggestions. For example, with transitive projects.

Even if we go that way, the only appropriate way of showing the installation suggestion (in my opinion) is a notification at the right bottom corner. We already have a similar one for a missing `dotnet-ef` tool:

![image](https://user-images.githubusercontent.com/20597871/186443351-95dd21cd-5493-473b-a55e-447e859f94b7.png)

Finally, there are some difficulties that I see even before starting implementing this:

- Need to track solution structure correctly. The plugin works in a reactive way, so we need to update the "candidates" collection on each solution- or project-level change
- Need to track the user's choice to not install the package if he clicks "Skip" or something on such notification. 
- Need to somehow configure which package (`.Tools` or `.Design`, they both work) should be automatically installed.
- Which version of a package we should install? The latest stable? The latest prerelease? If we go with the same version of what we have in a referenced Migrations project, which Migrations project to choose if we are referencing a couple of them? For example, for different database providers.

---

To summarize, it is possible to run this race, but I'm not sure about the effort. What about just having a comment right after the Startup project field, something like that:
> [Why](#) is my project not listed here?

And clicking on Why will open a help link with a detailed description.

**Reply by DiamondImpostor** [reply]
Yes, apparently the solution with "Why is my project not listed here?" is easier and more reliable, thanks for the answer.

Does it make sense to check Startup Project for \<none\>? Or check the number of projects on the startup projects list

I will create a PR with a help link for this case explaining what's best to do

**Reply by seclerp** [reply]
> Does it make sense to check Startup Project for `<none>`?

What do you mean? In the current implementation, validation logic doesn't allow you to execute commands without **Startup project** selection.

**Reply by DiamondImpostor** [reply]
I'm talking about whether or not to output "Why is my project not listed here?" under the startup project if the project is found. However, I think it is not so important

**Reply by seclerp** [reply]
Ah, I see. No, let's just show the comment permanently, as you may have other projects in your solution.

**Reply by seclerp** [reply]
@TheBottleCyber please use such URL as a hyperlink on a label:
```
https://plugins.jetbrains.com/plugin/18147-entity-framework-core-ui/f-a-q#why-i-cant-see-my-project-in-a-startup-projects-field
```

I will also create a feature issue in a couple of minutes.

**Edit**: Issue #114

**Reply by DiamondImpostor** [reply]
I just finished writing an .md file that you can put on github, I think it's the right thing to do. Should I create a pull request with the .md file?

**Reply by seclerp** [reply]
Lets move it to plugin marketplace F.A.Q. page. I will also add the link to this page to the README.

You could create a PR including your description and I will update the info on the F.A.Q. page.
